Story/cite 137

MaulikLim opened the pull request:
# Guidelines for Pull Requests

If you haven't yet read our code review guidelines, please do so, You can find them [here](https://diging.atlassian.net/wiki/spaces/DIGING/pages/2256076801/Code+Review+Guidelines).

Please confirm the following by adding an x for each item (turn `[ ]` into `[x]`).

- [x] I have removed all code style changes that are not necessary (e.g. changing blanks across the whole file that don’t need to be changed, adding empty lines in parts other than your own code)
- [x] I am not making any changes to files that don’t have any effect (e.g. imports added that don’t need to be added)
- [x] I do not have any sysout statements in my code or commented out code that isn’t needed anymore
- [x] I am not reformatting any files in the wrong format or without cause. 
- [x] I am not changing file encoding or line endings to something else than UTF-8, LF
- [x] My pull request does not show an insane amount of files being changed although my ticket only requires a few files being changed
- [x] I have added Javadoc/documentation where appropriate
- [x] I have added test cases where appropriate
- [x] I have explained any part of my code/implementation decisions that is not be self-explanatory

## Please provide a brief description of your ticket 
(you can copy the ticket if it hasn't changed) 

[CITE-137] - It should be possible to add new authorities
  
## Anything else the reviewer needs to know?

For the search functionality, most of the code was copied over from what is previously implemented. Appropriate changes and clean up are made on them to adapt it. For manually created authority, the generated URI is configurable through config properties.


[CITE-137]: https://diging.atlassian.net/browse/CITE-137

diging-jenkins commented:
Can one of the admins verify this patch?

diging-jenkins commented:
Can one of the admins verify this patch?

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/AuthorityService.java:
```diff
@@ -201,6 +203,14 @@ public IAuthorityEntry create(IAuthorityEntry entry, IUser user) {
+    public IAuthorityEntry create(IAuthorityEntry entry, IUser user, String uri) {
+        entry = create(entry, user);
+        entry.setUri(uri + entry.getId());
+        return save(entry);
```
have the other create method call this one and move the code from the other create method here; otherwise you'll end up with two save operations which is unnecessary.

MaulikLim replied:
I can merge both methods but wouldn't we require to save once to generate an Id so that we can append it to the generated URI. Also, we need to distinguish between the case where we need to generate a URI and where we don't. So, I created another method.

jdamerow replied:
jup, you're right. let's leave it the way it is but make a comment.

MaulikLim replied:
okay

jdamerow replied:
comment is still missing

MaulikLim replied:
I changed the method name, which I think is self explanatory. I'll add the comment for better clarity.

jdamerow replied:
the comment should be before or after `create` was called saying the it needs to be persisted first before we can get its id to create the uri.

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/authorities/AddAuthorityController.java:
```diff
@@ -0,0 +1,55 @@
+        entry.setName(form.getName());
+        entry.setDescription(form.getDescription());
+        entry.setGroups(new HashSet<>());
+        entry = authorityService.create(entry, (IUser) authentication.getPrincipal(), authorityUri + authorityPrefix);
```
uri creation should not be part of the controller; that's business logic

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/authorities/AddAuthorityController.java:
```diff
@@ -0,0 +1,55 @@
+        return "auth/authorities/new";
+    }
+
+    @PostMapping(value = "/auth/authority/new")
```
let's call that `create` instead of `new` (or whatever term is used for citations).

MaulikLim replied:
It is "add" for others, but there's already an "/auth/authority/add" POST api, so I used "new" to avoid confusion

jdamerow replied:
there is a method for `/auth/authority/create` but it actually imports an authority; let's change the existing one to `/auth/authority/import` and then make this one `/auth/authority/create`. it's more accurate.

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/authorities/AddAuthorityController.java:
```diff
@@ -0,0 +1,55 @@
+    @Autowired
+    private IAuthorityService authorityService;
+
+    @GetMapping(value = "/auth/authority/new")
```
see below

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/authorities/AuthorityEntryController.java:
```diff
@@ -63,9 +63,9 @@
     }
     
-    @RequestMapping("/auth/authority/{zoteroGroupId}/find/authorities/{source}")
+    @RequestMapping("/auth/authority/find/authorities/{source}")
```
instead of changing the url and risking missing a spot where it wouldn't work anymore, you could just add both paths to the request mapping (so that the method respond to both urls) and leave the existing code untouched

MaulikLim commented:
Addressed review comments - refactored some API end-points, renamed service method name

jdamerow commented:
Is there a reason this is still a draft?

MaulikLim commented:
I am not sure if I've addressed the scenarios appropriately. Mostly about the interactions in search pop-up. For e.g., showing results when searching by URI.

jdamerow commented on citesphere/src/test/java/edu/asu/diging/citesphere/core/service/impl/AuthorityServiceTest.java:
```diff
@@ -288,6 +291,36 @@ public AuthorityEntry answer(InvocationOnMock invocation) {
+            }
+        };
+
+        Thread.sleep(100);
```
why are you sleeping here?

jdamerow commented:
Make it so, Jenkins.

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented:
Make it so, Jenkins.

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented on citesphere/src/main/resources/config.properties:
```diff
@@ -4,6 +4,8 @@ _zotero_page_size=50
 _api_page_size=50
 _zotero_collections_max_number=30
+_authority_uri=${authority.uri}
+_authority_prefix=${authority.prefix}
```
these need to be added to the pom

jdamerow commented:
Make it so, Jenkins.

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented:
when I go to `auth/authority/list` I just get an empty page; no exception in the log.

MaulikLim commented:
For me it throws following error in logs. But when I rebuild the project with "mvn clean install" command, it starts working. I assumed it must be a dirty build issue. I did try to solve the following issue using several options, but it didn't work.

SEVERE: Servlet.service() for servlet [appServlet] in context with path [/citesphere] threw exception [Request processing failed; nested exception is org.thymeleaf.exceptions.TemplateInputException: An error happened during template parsing (template: "ServletContext resource [/WEB-INF/views/auth/authorities/list.html]")] with root cause
org.thymeleaf.exceptions.TemplateProcessingException: Could not parse as expression: "https://localhost:8080/citesphere/auth/authority/" (template: "auth/authorities/list" - line 90, col 12)
	at org.thymeleaf.standard.expression.StandardExpressionParser.parseExpression(StandardExpressionParser.java:131)
	at org.thymeleaf.standard.expression.StandardExpressionParser.parseExpression(StandardExpressionParser.java:62)
	at org.thymeleaf.standard.expression.StandardExpressionParser.parseExpression(StandardExpressionParser.java:44)

Once I've cleaned it, it loads the page correctly.

MaulikLim commented:
Authority list page is working fine now.

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/core/service/impl/AuthorityService.java:
```diff
@@ -201,6 +210,15 @@ public IAuthorityEntry create(IAuthorityEntry entry, IUser user) {
 
+    @Override
+    @Transactional
+    //Creates a new Authority entry and also generates a custom URI for it
```
needs proper javadoc

jdamerow commented on citesphere/src/test/java/edu/asu/diging/citesphere/core/service/impl/AuthorityServiceTest.java:
```diff
@@ -288,6 +291,34 @@ public AuthorityEntry answer(InvocationOnMock invocation) {
     }
+    
+    @Test
+    public void test_createWithURI() throws InterruptedException {
```
when is that exception being thrown?

jdamerow reviewed (changes requested):
Is there a reason this is still a draft PR?

MaulikLim commented:
Added javadoc for createWithURI() method, added proper comment and removed unnecessary thrown exception

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented on citesphere/src/main/webapp/WEB-INF/views/auth/authorities/create.html:
```diff
@@ -0,0 +1,375 @@
+
+    $("#authorIconContainer").on('click', ".popover #authorCreateAuthority", function() {
+        var uri = $("#uriLoadingFoundAuthor").attr('data-authority-uri');
+        $.post([[@{|/auth/authority/import?${_csrf.parameterName}=${_csrf.token}&uri=|}]]+uri, function(data) {
```
csrf token should be passed in body of request

jdamerow commented on citesphere/src/main/webapp/WEB-INF/views/auth/authorities/create.html:
```diff
@@ -0,0 +1,375 @@
+    }
+}
+
+function getconceptpowerAuthorities(modalType, personType, page) {
```
should be "get**C**onc..."

jdamerow commented:
Make it so, Jenkins.

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented:
- if an authority is imported through the modal, the new authority should get the URI of the imported authority (like when you import an authority via the edit item page)
- there needs to be some explanation about what the search button does (e.g. add something like "To import an authority from another service, click the search button."

MaulikLim commented:
Addressed review comments: Passed csrf token through body(also made the same correction for editItem.html), renamed method, added instruction for search button

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented:
Make it so, Jenkins.

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented:
For importing, the flow needs to be a little bit changed (just for importing from the create authority view):
- If the user chooses to search for an authority instead of creating a new one, they should see the search screen and once they selected on, the fields of the create authority page should be filled.
- There should be a message that tells the user what URI will be used.
- Only when the user clicks create, the authority would be created.

- Also, there needs to be a dropdown that lets the user select which group to add the new authority to.
- And there is a bug. If I select a group on the list page, then the add button doesn't work anymore.
- If the add button is clicked after a group has been selected, then the group should be preselected in the create authority form.

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented on citesphere/src/main/java/edu/asu/diging/citesphere/web/user/authorities/CreateAuthorityController.java:
```diff
@@ -0,0 +1,79 @@
+        IUser user = (IUser) authentication.getPrincipal();
+        model.addAttribute("groups", citationManager.getGroups(user));
+        CreateAuthorityForm authorityForm = new CreateAuthorityForm();
+        if (zoteroGroupId != null) {
```
you can probably skip this. If `zoteroGroupId` is null, it will just replace null with null.

jdamerow commented:
Make it so, Jenkins.

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented:
After a new authority has been created in a group, the authority list page should be filtered by that group.

Besides that (and my comment on the code), should be good to go.

MaulikLim commented:
Addressed review comments - removed unwanted null check, filtered authority list page by groupId when an authority is created in a group

diging-jenkins commented:
Congratulations, build succeeded!

jdamerow commented:
Make it so, Jenkins.

diging-jenkins commented:
Congratulations, build succeeded!
